shared/src/main/java/com/couchbase/lite/LiveQuery.java

```diff
@@ -0,0 +1,172 @@
+package com.couchbase.lite;
+
+import android.os.Handler;
+import android.os.Looper;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A Query subclass that automatically refreshes the result rows every time the database changes.
+ */
+public class LiveQuery implements DatabaseChangeListener {
+
+    //---------------------------------------------
+    // static variables
+    //---------------------------------------------
+    private final static String TAG = Log.QUERY;
+    private final static long kDefaultLiveQueryUpdateInterval = 200; // 0.2sec (200ms)
+
+    //---------------------------------------------
+    // member variables
+    //---------------------------------------------
+
+    private Set<LiveQueryChangeListener> queryChangeListener;
+    private Query query;
+    private ResultSet resultSet;
+    private boolean observing;
+    private boolean willUpdate;
+    private boolean forceReload;
+    private long lastUpdatedAt;
+
+    //---------------------------------------------
+    // Constructors
+    //---------------------------------------------
+
+    /* package */ LiveQuery(Query query) {
+        this.query = query;
+        queryChangeListener = Collections.synchronizedSet(new HashSet<LiveQueryChangeListener>());
+        resultSet = null;
+        observing = false;
+        willUpdate = false;
+        forceReload = false;
+        lastUpdatedAt = 0L;
+    }
+
+    //---------------------------------------------
+    // API - public methods
+    //---------------------------------------------
+
+    /**
+     * Starts observing database changes. The .rows property will now update automatically.
+     * (You usually don't need to call this yourself, since accessing or observing the getRows()
+     * property will call -start for you.)
+     */
+    public void run() {
+        if (!observing) {
+            observing = true;
+            query.getDatabase().addChangeListener(this);
+            update();
+        }
+    }
+
+    /**
+     * Stops observing database changes. Calling start() or getRows() will restart it.
+     */
+    public void stop() {
+        if (observing) {
+            observing = false;
+            query.getDatabase().removeChangeListener(this);
+        }
+        willUpdate = false; // cancels the delayed update started by -databaseChanged
+    }
+
+    public void addChangeListener(LiveQueryChangeListener listener) {
+        queryChangeListener.add(listener);
+        if (queryChangeListener.size() == 1)
+            run();
```

Comment:
This implementation does not work well with the behavior required of `run()` method.

If we add this, we should change `run()` method as a non-public method.